src/detect-ftp-mode.c

```diff
@@ -0,0 +1,186 @@
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+/**
+ *
+ * \author Jeff Lucovsky <jlucovsky@oisf.net>
+ *
+ * Implements the ftp.mode sticky buffer
+ *
+ */
+
+#include "suricata-common.h"
+
+#include "detect.h"
+#include "detect-parse.h"
+#include "detect-engine.h"
+
+#include "flow.h"
+
+#include "util-debug.h"
+
+#include "app-layer.h"
+#include "app-layer-ftp.h"
+
+#include "detect-ftp-mode.h"
+
+#define KEYWORD_NAME "ftp.mode"
+#define KEYWORD_DOC  "ftp-keywords.html#ftp-mode"
+#define BUFFER_NAME  "ftp.mode"
+#define BUFFER_DESC  "ftp mode"
+
+typedef struct DetectFtpModeData_ {
+    bool active;
+} DetectFtpModeData;
+
+/**
+ * \brief Regex for parsing our keyword options
+ */
+#define PARSE_REGEX "^\\s*(active|passive)\\s*$"
+static DetectParseRegex parse_regex;
+
+static int g_ftp_mode_buffer_id = 0;
+
+/**
+ * \brief This function is used to check matches from the FTP App Layer Parser
+ *
+ * \param t pointer to thread vars
+ * \param det_ctx pointer to the pattern matcher thread
+ * \param p pointer to the current packet
+ * \param m pointer to the sigmatch
+ * \retval 0 no match
+ * \retval 1 match
+ */
+static int DetectFtpModeMatch(DetectEngineThreadCtx *det_ctx, Flow *f, uint8_t flags, void *state,
+        void *txv, const Signature *s, const SigMatchCtx *m)
+{
+    FTPTransaction *tx = (FTPTransaction *)txv;
+    if (tx->command_descriptor.command_code == FTP_COMMAND_UNKNOWN) {
+        return 0;
+    }
+    if (!tx->dyn_port) {
+        return 0;
+    }
+
+    const DetectFtpModeData *ftpmoded = (const DetectFtpModeData *)m;
+    return 1 ? ftpmoded->active == tx->active : 0;
+}
+
+/**
+ * \brief this function will free memory associated with DetectFtpModeData
+ *
+ * \param ptr pointer to DetectFtpModeData
+ */
+static void DetectFtpModeFree(DetectEngineCtx *de_ctx, void *ptr)
+{
+    SCFree(ptr);
+}
+
+/**
+ * \brief This function is used to parse ftp.mode options passed via ftp.mode keyword
+ *
+ * \param str Pointer to the user provided ftp.mode options
+ *
+ * \retval  pointer to DetectFtpModeData on success
+ * \retval NULL on failure
+ */
+static DetectFtpModeData *DetectFtpModeParse(const char *optstr)
+{
+    DetectFtpModeData *ftpmoded = NULL;
+    char arg1[8] = "";
+    size_t pcre2len;
+    pcre2_match_data *match = NULL;
+
+    int ret = DetectParsePcreExec(&parse_regex, &match, optstr, 0, 0);
```

Comment:
do we really need all this pcre logic to simply parse a single arg? We're generally trying to get away from using pcre for parsing, so lets not use it if we can avoid it

Reply:
Yes ...  I only need a trimmed string so I'll add a function to parser.rs and use its result.